### DISCUSSION TITLE
Realtime + RLS, not working on hosted supabase instance.

### DISCUSSION THREAD

**Discussion by skizer** [supabase/supabase#7630] [Questions]
Hey folks,

after reading tons of discord messages and old issues on github it seems like there's something wrong with the hosted instance of supabase in correlation with Realtime + RLS.

Access is all granted via policy, socket gets created and connects and the realtime->subscription table gets populated with the subscription.
According to the latest blog post, the hosted version of supabase supports RLS+Realtime, am I wrong?

~~I've struggled a bit with a fairly simple scenario for a while now until I found the official blog post where they've announced that RLS now also affects the real-time subscription feature.~~

~~So with the RLS in effect for subscriptions, is it even possible now to notify subscribers (anon for example) on things like DELETE and UPDATE even tho they only have permission for SELECT?~~

**Comment by GaryAustin1** [comment]
Unless it broke in the last couple of days, and I’ve seen no reports of that (assuming you also asked this same thing on discord) , it works with RLS on hosted.  

One thing you could check is if it sends a user or anon token in the websocket message.  
How are you subscribing to realtime on the client?

**Reply by GaryAustin1** [reply]
Also double check you are running on a recent client software version.  Ran into someone who had supabase-js from last year and it does not support the RLS feature.

**Reply by skizer** [reply]
Just to be on the safe side, I've cleared all npm caches and setup a fresh vite/vue project with the latest supabase.js client version and tested it again.

**Comment by skizer** [comment]
Hey gary, thanks for your response!

I do really have no clue why it's not working, I've checked the websocket, it sends the anon token as user_token, the server responses correctly, the subscription get's added to the database with the anon role set.

I can do CRUD operations just fine with the anon token for the desired table in the public schema.
All I know is the moment I activate RLS for that table that realtime doesn't fire anymore.

```
onMounted(() => {
  sub.value = client
    .from("test")
    .on("*", (payload) => {
      console.log("updated", payload);
    })
    .subscribe();
});
```

**Reply by GaryAustin1** [reply]
This is what the websocket should look like for a RLS update (select policy role authenticated) with a signed in user:  Note the update message.
<img width="391" alt="update" src="https://user-images.githubusercontent.com/54564956/178114989-5bb29a5f-d4b8-461e-88bd-48ec8e1d6eb6.PNG">
This is what socket looks like with no user logged in.  Note the access token is null:
<img width="400" alt="nouser" src="https://user-images.githubusercontent.com/54564956/178115009-23a74128-b423-45a3-a780-2270f28652e4.PNG">
Note the difference in having an access token for logged in user versus not above.  The access token is filled in with the user jwt in follow up heartbeats.  Logged in user:
<img width="455" alt="user" src="https://user-images.githubusercontent.com/54564956/178115051-bb3d6dcf-f62e-4890-b95d-affbf997c4d1.PNG">

**Reply by skizer** [reply]
Wait, does that mean it won't work with only the anon_public token set when init. the supabase.js client?
Like that's literally what they've said, the changes should make it work with the anon role too.

Btw. I've tried it with faking the auth by setting supabase.auth.setAuth to the anon_token, so the access_token gets filled with the anon_token. Still won't fire.

May I ask if this example you've posted is from a fresh project?

![image](https://user-images.githubusercontent.com/3046198/178115780-2c7d1e60-38d9-4b65-b171-6021ea7a21f2.png)

P.S https://github.com/supabase/supabase/discussions/4342

//Update
Just to make sure, it also doesn't work with authenticated users/role

**Reply by GaryAustin1** [reply]
Anon works fine too with no user, if your select RLS allows it.  I think this test instance I'm using is 6 months old, but my year old one worked with RLS last time I checked.

**Reply by GaryAustin1** [reply]
With no logged in user, RLS on, select policy for anon.
![image](https://user-images.githubusercontent.com/54564956/178117208-68910585-bde3-4780-8f8c-2303c239a7d5.png)

**Comment by joshua1996** [comment]
Realtime + RLS only working in DELETE event

**Reply by w3b6x9** [reply]
@joshua1996 is this still the case for you?

**Comment by m0nac0** [comment]
I have a similar issue with the Flutter SDK. No matter what I try, as soon as I enable RLS (even with the 'Enable read access for all users' template policy enabled) no realtime updates are received anymore. I can still see the authenticated subscription in the `realtime.subscription` table, though.
And when the app executes a SELECT, it receives all the new rows, so it's not an issue with the RLS policies.

As soon as I disable RLS, realtime starts working again.

**Reply by GaryAustin1** [reply]
There was an issue on Supabase side, fix being rolled out now. https://github.com/supabase/realtime/issues/265#issuecomment-1181560093

**Reply by m0nac0** [reply]
Thanks, that's good to hear. I started to think I'm stupid, since my code is functionally equivalent to the official examples and I still didn't get it to work in the last 2 days 😂

**Reply by m0nac0** [reply]
I can confirm that running the query linked above instantly fixed the issue for me.

**Comment by w3b6x9** [comment]
@skizer are you still having Realtime issues?

**Comment by ankitparmar007** [comment]
@w0wka91 I am facing same problem

**Reply by LivioGama** [reply]
Same for me, using a policy with custom auth provider and jwt

**Reply by LivioGama** [reply]
This can also be due to this if you have trouble with realtime, upgrade supabase client to 2.24.0 https://github.com/supabase/supabase-js/commit/b55d386b35cdf9373fabbbc07199f86811bde339

**Reply by LivioGama** [reply]
That was not it, I decided to write a whole medium post for avoiding you waste time in the future: https://liviogamassia.medium.com/using-supabase-rls-with-firebase-auth-custom-auth-provider-357eaad9c70f

**Comment by huyouare** [comment]
I was also having an issue with RLS not working as expected when combined with the Realtime API on a Supabase hosted instance. Adding this line fixed the problem for me:

```
const supabaseAccessToken = await getToken({ template: "supabase" });
const supabase = await supabaseClient(supabaseAccessToken as string);
supabase.realtime.accessToken = supabaseAccessToken; // THIS IS REQUIRED FOR RLS!!!
```

**Reply by simonhackler** [reply]
Insane. Thank you so much. Sometimes this token would be set automatically for me and sometimes not. This fixed it.

**Reply by mustafakendiguzel** [reply]
not worked on me

**Reply by t0mdicks0n** [reply]
Thank you so much for sharing this tip! I was struggling with real-time subscriptions in Supabase where my RLS policies worked for regular queries but didn’t apply correctly for WebSocket connections. I could see DELETE events but not INSERT or UPDATE events, and everything worked perfectly when I temporarily disabled RLS. After hours of debugging and testing different RLS configurations, I couldn’t figure out what was wrong.

Your suggestion to explicitly set the access token using supabase.realtime.setAuth() was exactly what I needed. It turns out that while my client was authenticated for API calls, the WebSocket connection wasn’t passing the token for RLS evaluation. By explicitly setting the token, I was able to fix the issue, and now all real-time events (INSERT, UPDATE, and DELETE) are working as expected.

**Comment by ddematheu** [comment]
@huyouare just came here to say that you are a gentleman and a scholar, this fixed it for me.
